### DISCUSSION TITLE
Spacy 3.0 merge_entities not merging entities when predicting from trained custom ner model(trained new model on conll data using CLI)

### DISCUSSION THREAD

**Discussion by Ashufet** [explosion/spaCy#7753] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

```python
import pandas as pd
from tqdm import tqdm
import spacy
from spacy.tokens import DocBin
import json
from datetime import datetime

nlp = spacy.load("./output/model-best") # ner model trained on conll data
print(nlp.pipe_names)
# ['tok2vec', 'ner']

nlp.add_pipe("merge_entities") # added merge_entities pipeline

print(nlp.pipe_names)
# ['tok2vec', 'ner', 'merge_entities']

texts = ["I live in New York USA",
        "Steve Smith is a cricketer"]

results = []
index = 0

print("Starting Spacy Ner")
now = datetime.now()
end_count = 1
n = 0

while n < end_count:
    n += 1
    for doc in nlp.pipe(texts):
        index += 1
        for ent in doc.ents:
            temp = [index,ent.label_,ent.text,ent.start_char,ent.end_char]
            results.append(temp)

elapsed_time = datetime.now() - now
print("Ner done",elapsed_time )
```

results
```
[[1, 'I-LOC', 'New', 10, 13],
 [1, 'I-LOC', 'York', 14, 18],
 [1, 'I-LOC', 'USA', 19, 22],
 [2, 'I-PER', 'Steve', 0, 5],
 [2, 'I-PER', 'Smith', 6, 11]]
```

I am expecting that the New York and Steve Smith should come as one word because we have used merge_entities in the pipeline.
Please let me know if I am doing anything worng.

## Your Environment
Spacy 3.0.5

**Comment by adrianeboyd** [comment, marked as answer]
Hi, it looks like your training data wasn't in the right format. The labels (`ent.label_`) should be `LOC` and `PER`, without the `IOB` part of the tag. Your model seems to have learned the individual IOB tags with tokens as the spans instead of learning the entities as whole spans, which explains why you don't see any merging: it thinks each token is a separate entity.

**Reply by Ashufet** [reply]
Thanks, @adrianeboyd for the suggestion. I trained the model again by removing IOB tags but still getting entities as different

Train data sample -
('West Indian allrounder Phil Simmons took four for 38 on Friday as Leicestershire beat Somerset by an innings and 39 runs in two days to take over at the head of the county championship .',
  {'entities': [(0, 4, 'MISC'),
    (5, 11, 'MISC'),
    (23, 27, 'PERSON'),
    (28, 35, 'PERSON'),
    (66, 80, 'ORG'),
    (86, 94, 'ORG')]}),

Test data output -

[[1, 'LOC', 'New', 10, 13],
 [1, 'LOC', 'York', 14, 18],
 [1, 'LOC', 'USA', 19, 22],
 [2, 'PERSON', 'Steve', 0, 5],
 [2, 'PERSON', 'Smith', 6, 11]]

**Reply by adrianeboyd** [reply]
You still have the entity spans covering single tokens rather than full entity spans. The training data should have a single entry for each entity span, so the whole span defined by sequence of the B-TAG + I-TAGs, e.g.:

```
(0, 11, 'MISC'), # West Indian
(23, 35, 'PERSON'), # Phil Simmons
...
```

Does that make sense?

**Reply by Ashufet** [reply]
Thanks will update the training and train the model again.

But I have one question, if we are training complete entry as one token and entity then what's the use of merge_entity pipeline because the model will predict the complete entry as one as it has learned this way.

**Reply by adrianeboyd** [reply]
The entity prediction (provided as spans in `doc.ents`) does not modify the tokenization of the doc (inspect `[t.text for t in doc]`), so `John` and `Smith` would still be two separate tokens even though there's an entity `John Smith`.

Using `merge_entities` will modify the tokens so that `John Smith` is one token in the doc instead.

**Reply by Ashufet** [reply]
Thanks for the info. It worked!!

**Reply by Sumit5194** [reply]
hello I am facing same problem. My ner model output is like United B-geo states I-geo but i want output should United States geo only. I gave training data eg- {"text": "There is United States, "entities":[(8 , 14, B-geo),(16,21,I-geo)]} like wise?

**Reply by adrianeboyd** [reply]
There are two supported formats for entity tags.

One is text + character spans and the other is text + words + IOB tags. The character spans should be for the whole entity span and do not include `B-` or `I-`.

```python
import spacy
from spacy.tokens import Doc
from spacy.training import Example

nlp = spacy.blank("en")

text = "There is United States"

ann1 = {
    "entities": [(9, 22, "geo")],
}
ex1 = Example.from_dict(nlp.make_doc(text), ann1)
print(ex1.reference.ents)

words2 = ["There", "is", "United", "States"]
ents2 = ["O", "O", "B-geo", "I-geo"]
doc2 = Doc(nlp.vocab, words=words2, ents=ents2)
ex2 = Example(nlp.make_doc(text), doc2)
print(ex2.reference.ents)
```

If you're still running into problems, please open a new thread with a full example showing what you're trying to do.
